rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+```
+Proposed:
+```python
+x1 = tf.keras.Input(shape=(1,), name=key_1, dtype=dtype)
+x2 = tf.keras.Input(shape=(1,), name=key_2, dtype=dtype)
+layer1 = tf.keras.layers.CategoryLookup(
+           vocabulary=vocabulary_list,  
+           num_oov_tokens=num_oov_buckets)
+x1 = layer1(x1)
+layer2 = tf.keras.layers.CategoryHashing(
+           num_bins=hash_bucket_size)
+x2 = layer2(x2)
+layer = tf.keras.layers.CategoryCrossing(num_bins=hash_bucket_size)
+out = layer([x1, x2])
+```
+
+6. weighted categorical column
+
+Original:
+```python
+fc = tf.feature_column.categorical_column_with_vocab_list(key, vocabulary_list,
+         dtype, default_value, num_oov_buckets)
+weight_fc = tf.feature_column.weighted_categorical_column(fc, weight_feature_key, 
+         dtype=weight_dtype)
+linear_model = tf.estimator.LinearClassifier(units, feature_columns=[weight_fc])
+```
+Proposed:
+```python
+x1 = tf.keras.Input(shape=(1,), name=key, dtype=dtype)
+x2 = tf.keras.Input(shape=(1,), name=weight_feature_key, dtype=weight_dtype)
+layer = tf.keras.layers.CategoryLookup(
+           vocabulary=vocabulary_list,   
+           num_oov_tokens=num_oov_buckets)
+x1 = layer(x1)
+x = tf.keras.layers.CategoryEncoding(num_categories=len(vocabulary_list)+num_oov_buckets)([x1, x2])
+linear_model = tf.keras.premade.LinearModel(units)
+linear_logits = linear_model(x)
+```
+
+## Design Proposal
+We propose a CategoryLookup layer to replace `tf.feature_column.categorical_column_with_vocabulary_list` and `tf.feature_column.categorical_column_with_vocabulary_file`, a `CategoryHashing` layer to replace `tf.feature_column.categorical_column_with_hash_bucket`, a `CategoryCrossing` layer to replace `tf.feature_column.crossed_column`, and another `CategoryEncoding` layer to convert the sparse input to the format required by linear models.
+
+```python
+`tf.keras.layers.CategoryLookup`
+CategoryLookup(PreprocessingLayer):
+"""This layer transforms categorical inputs to index space.
+   If input is dense/sparse, then output is dense/sparse."""
+
+  def __init__(self, max_tokens=None, num_oov_tokens=1, vocabulary=None,
+               name=None, **kwargs):
+    """Constructs a CategoryLookup layer.
+
+    Args:
+      max_tokens: The maximum size of the vocabulary for this layer. If None,
+              there is no cap on the size of the vocabulary. This is used when `adapt`
+              is called.
+      num_oov_tokens: Non-negative integer. The number of out-of-vocab tokens. 
+              All out-of-vocab inputs will be assigned IDs in the range of 
+              [0, num_oov_tokens) based on a hash. When
+              `vocabulary` is None, it will convert inputs in [0, num_oov_tokens)
+      vocabulary: the vocabulary to lookup the input. If it is a file, it represents the 
+              source vocab file; If it is a list/tuple, it represents the source vocab 
+              list; If it is None, the vocabulary can later be set.
```

Comment:
what is the format of the file? how do you set the vocabulary later? what is the expected use of the `adapt` method?

Reply:
1) format of the file is same as 
a) any other TFX vocab file, or 
b) this test file: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/feature_column/testdata/warriors_vocabulary.txt

2) users from feature columns world will set it during __init__, but this layer also allow users to call 'adapt' to derive/set the vocabulary from dataset.

Reply:
My point is that this should be documented. Stating the the vocabulary can be set later without showing how is not useful.

Reply:
Done.